README.md
# OneWaySMS SDK for Go

A non-official OneWaySMS SDK written for Go. Based on the latest [v1.3](http://smsd2.onewaysms.sg/api.pdf) OneWaySMS API documentation.

[![license](https://img.shields.io/github/license/junwen-k/onewaysms-sdk-go)](https://raw.githubusercontent.com/junwen-k/onewaysms-sdk-go/master/LICENSE.txt)

## Installing

Use `go get` to retrieve the SDK to add it to your `GOPATH` workspace, or project's Go module dependencies.

    go get github.com/junwen-k/onewaysms-sdk-go

To update the SDK use `go get -u` to retrieve the latest version of the SDK.

    go get -u github.com/junwen-k/onewaysms-sdk-go

## Dependencies

The SDK includes a `vendor` folder containing the runtime dependencies of the SDK. The metadata of the SDK's dependencies can be found in the Go module file `go.mod`.

## Usage and Getting Started

### Importing the SDK

Include the following import statement to use the SDK.

```go
package main

import (
  "github.com/junwen-k/onewaysms-sdk-go/owerr"
  "github.com/junwen-k/onewaysms-sdk-go/owsms"
)
```

### Initializing a new client

Initialize a new client using the `NewClient` function. For instance:

```go
func main() {
	svc := owsms.NewClient(
		"API_BASE_URL",
		"API_USERNAME",
		"API_PASSWORD",
		"SENDER_ID",
  )
  // ...
}
```

Optionally, a new client with a custom HTTP client can be initialized using the `NewClientWithHTTP` function. For instance:

```go
func main() {
	svc := owsms.NewClientWithHTTP(
		"API_BASE_URL",
		"API_USERNAME",
		"API_PASSWORD",
		"SENDER_ID",
		&http.Client{Timeout: time.Second * 30},
  )
  // ...
}
```

### Use case examples

1. **Send SMS** - Send SMS by calling OneWaySMS API gateway, returning mobile terminating ID(s) if request is successful.

   ```go
    func main() {
      // ...
      output, _, err := svc.SendSMS(&owsms.SendSMSInput{
        Message:  "Hello, 世界",
        MobileNo: []string{"60123456789", "60129876543"},
      })
      if err != nil {
        if owErr, ok := err.(owerr.Error); ok {
          switch owErr.Code() {
          case owerr.RequestFailure:
          // Handle RequestFailure
          case owerr.InvalidCredentials:
          // Handle InvalidCredentials
          case owerr.InvalidSenderID:
          // Handle InvalidSenderID
          case owerr.InvalidMobileNo:
          // Handle InvalidMobileNo
          case owerr.InvalidLanguageType:
          // Handle InvalidLanguageType
          case owerr.InvalidMessageCharacters:
          // Handle InvalidMessageCharacters
          case owerr.InsufficientCreditBalance:
          // Handle InsufficientCreditBalance
          case owerr.UnknownError:
            // Handle UnknownError
          default:
          }
        } else {
          // Handle Generic Error
        }
      }

      // MTIDs - Mobile terminating IDs
      fmt.Println(output.MTIDs)
    }
   ```

1. **Check MT Transaction Status** - Check mobile terminating transaction status based on mobile terminating ID provided. Mobile terminating ID can be obtained by calling send SMS API.

   ```go
    func main() {
      // ...
      output, _, err := svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
        MTID: 145712470,
      })
      if err != nil {
        if owErr, ok := err.(owerr.Error); ok {
          switch owErr.Code() {
          case owerr.MTInvalidNotFound:
          // Handle MTInvalidNotFound
          case owerr.MessageDeliveryFailure:
          // Handle MessageDeliveryFailure
          case owerr.UnknownError:
            // Handle UnknownError
          default:
          }
        } else {
          // Handle Generic Error
        }
      }

      switch output.Status {
      case owsms.MTTransactionStatusSuccess:
        // Handle success status
      case owsms.MTTransactionStatusTelcoDelivered:
        // Handle telco delivered status
      default:
      }
    }
   ```

1. **Check Credit Balance**. Check remaining credit balance for the account in the client's config.

   ```go
    func main() {
      // ...
      output, _, err := svc.CheckCreditBalance()
      if err != nil {
        if owErr, ok := err.(owerr.Error); ok {
          switch owErr.Code() {
          case owerr.InvalidCredentials:
          // Handle InvalidCredentials
          case owerr.UnknownError:
            // Handle UnknownError
          default:
          }
        } else {
          // Handle Generic Error
        }
      }

      // CreditBalance - Remaining credit balance for this account
      fmt.Println(output.CreditBalance)
    }
   ```

## License

This SDK is distributed under the MIT License, see LICENSE.txt for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2020-06-03

- Initial release


owsms/types_test.go
// Copyright (c) KwanJunWen
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

package owsms_test

import (
	"testing"

	"github.com/junwen-k/onewaysms-sdk-go/owsms"
	"github.com/pkg/errors"
	"github.com/stretchr/testify/assert"
)

func TestSendSMSInputValidate(t *testing.T) {
	tests := []struct {
		desc     string
		input    *owsms.SendSMSInput
		expected error
	}{
		{
			desc: "With valid values",
			input: &owsms.SendSMSInput{
				LanguageType: owsms.LanguageTypeNormal,
				Message:      "Hello World",
				MobileNo:     []string{"60123456789"},
			},
			expected: nil,
		},
		{
			desc: "With invalid LanguageType",
			input: &owsms.SendSMSInput{
				LanguageType: "invalid",
				Message:      "Hello World",
				MobileNo:     []string{"60123456789"},
			},
			expected: errors.New("SendSMSInput: Error: LanguageType is invalid"),
		},
		{
			desc: "With missing Message",
			input: &owsms.SendSMSInput{
				LanguageType: owsms.LanguageTypeNormal,
				Message:      "",
				MobileNo:     []string{"60123456789"},
			},
			expected: errors.New("SendSMSInput: Error: Message is required"),
		},
		{
			desc: "With missing MobileNo",
			input: &owsms.SendSMSInput{
				LanguageType: owsms.LanguageTypeNormal,
				Message:      "Hello World",
				MobileNo:     nil,
			},
			expected: errors.New("SendSMSInput: Error: MobileNo is required"),
		},
	}
	for _, test := range tests {
		t.Run(test.desc, func(t *testing.T) {
			actual := test.input.Validate()
			if actual != nil {
				assert.EqualError(t, test.expected, actual.Error())
			} else {
				assert.Equal(t, test.expected, actual)
			}
		})
	}
}

func TestCheckTransactionStatusInputValidate(t *testing.T) {
	tests := []struct {
		desc     string
		input    *owsms.CheckTransactionStatusInput
		expected error
	}{
		{
			desc: "With valid values",
			input: &owsms.CheckTransactionStatusInput{
				MTID: 145712468,
			},
			expected: nil,
		},
		{
			desc: "With missing MTID",
			input: &owsms.CheckTransactionStatusInput{
				MTID: 0,
			},
			expected: errors.New("CheckTransactionStatusInput: Error: MTID is required"),
		},
	}
	for _, test := range tests {
		t.Run(test.desc, func(t *testing.T) {
			actual := test.input.Validate()
			if actual != nil {
				assert.EqualError(t, test.expected, actual.Error())
			} else {
				assert.Equal(t, test.expected, actual)
			}
		})
	}
}


owsms/client_test.go
// Copyright (c) KwanJunWen
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

package owsms_test

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/junwen-k/onewaysms-sdk-go/owerr"
	"github.com/junwen-k/onewaysms-sdk-go/owsms"
	"github.com/stretchr/testify/assert"
)

func TestSendSMS(t *testing.T) {
	var (
		svc    *owsms.Client
		output *owsms.SendSMSOutput
		err    error
	)

	t.Run("With valid values", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
			fmt.Fprintln(w, "145712468")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "Hello World",
			MobileNo: []string{"60123456789"},
		})
		assert.NoError(t, err)
		assert.NotNil(t, output)
		assert.Equal(t, []int{145712468}, output.MTIDs)
	})

	t.Run("With multiple mobileNo", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
			fmt.Fprintln(w, "145712468,145712469")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "Hello World",
			MobileNo: []string{"60123456789", "60129876543"},
		})
		assert.NoError(t, err)
		assert.NotNil(t, output)
		assert.Equal(t, []int{145712468, 145712469}, output.MTIDs)
	})

	t.Run("With request failure", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusInternalServerError)
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "request failure",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 500 (Internal Server Error): request failure")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "request failure", owErr.Message())
		assert.Equal(t, owerr.RequestFailure, owErr.Code())
		assert.Equal(t, http.StatusInternalServerError, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With invalid user credentials", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-100")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "invalid", "invalid", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "Hello World",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): apiusername or apipassword is invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "apiusername or apipassword is invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidCredentials, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With invalid senderID", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-200")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "invalid")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "Hello World",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): senderid parameter is invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "senderid parameter is invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidSenderID, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With invalid mobileNo", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-300")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "Hello World",
			MobileNo: []string{"invalid"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): mobileno parameter is invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "mobileno parameter is invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidMobileNo, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With invalid languageType", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-400")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:      "Hello World",
			MobileNo:     []string{"60123456789"},
			LanguageType: "invalid",
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): languagetype is invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "languagetype is invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidLanguageType, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With invalid message", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-500")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "invalid",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): characters in message are invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "characters in message are invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidMessageCharacters, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With insufficient credit balance", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-600")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "insufficient credit balance",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): insufficient credit balance")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "insufficient credit balance", owErr.Message())
		assert.Equal(t, owerr.InsufficientCreditBalance, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With unknown error", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "random")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.SendSMS(&owsms.SendSMSInput{
			Message:  "unknown error",
			MobileNo: []string{"60123456789"},
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): unknown error")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "unknown error", owErr.Message())
		assert.Equal(t, owerr.UnknownError, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})
}

func TestCheckTransactionStatus(t *testing.T) {
	var (
		svc    *owsms.Client
		output *owsms.CheckTransactionStatusOutput
		err    error
	)

	t.Run("With success status", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "0")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
			MTID: 145712470,
		})
		assert.NoError(t, err)
		assert.NotNil(t, output)
		assert.Equal(t, owsms.MTTransactionStatusSuccess, output.Status)
	})

	t.Run("With telco_delivered status", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "100")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
			MTID: 145712471,
		})
		assert.NoError(t, err)
		assert.NotNil(t, output)
		assert.Equal(t, owsms.MTTransactionStatusTelcoDelivered, output.Status)
	})

	t.Run("With invalid mtID", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-100")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
			MTID: 1,
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): mtid is invalid or not found")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "mtid is invalid or not found", owErr.Message())
		assert.Equal(t, owerr.MTInvalidNotFound, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With failed mtID", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-200")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
			MTID: 1,
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): message delivery failed")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "message delivery failed", owErr.Message())
		assert.Equal(t, owerr.MessageDeliveryFailure, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With unknown error", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "random")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckTransactionStatus(&owsms.CheckTransactionStatusInput{
			MTID: 1,
		})
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): unknown error")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "unknown error", owErr.Message())
		assert.Equal(t, owerr.UnknownError, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})
}

func TestCheckCreditBalance(t *testing.T) {
	var (
		svc    *owsms.Client
		output *owsms.CheckCreditBalanceOutput
		err    error
	)

	t.Run("With success status", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "6500.5")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "Username", "Password", "SenderID")

		output, _, err = svc.CheckCreditBalance()
		assert.NoError(t, err)
		assert.NotNil(t, output)
		assert.Equal(t, float32(6500.5), output.CreditBalance)
	})

	t.Run("With invalid user credentials", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "-100")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "invalid", "invalid", "SenderID")

		output, _, err = svc.CheckCreditBalance()
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): apiusername or apipassword is invalid")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "apiusername or apipassword is invalid", owErr.Message())
		assert.Equal(t, owerr.InvalidCredentials, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})

	t.Run("With unknown error", func(t *testing.T) {
		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintln(w, "random")
		}))
		defer ts.Close()

		svc = owsms.NewClient(ts.URL, "invalid", "invalid", "SenderID")

		output, _, err = svc.CheckCreditBalance()
		assert.Error(t, err)
		assert.EqualError(t, err, "OneWaySMS: Error 200 (OK): unknown error")
		owErr, ok := err.(owerr.Error)
		assert.True(t, ok)
		assert.Equal(t, "unknown error", owErr.Message())
		assert.Equal(t, owerr.UnknownError, owErr.Code())
		assert.Equal(t, http.StatusOK, owErr.StatusCode())
		assert.Nil(t, output)
	})
}


owerr/error.go
// Copyright (c) KwanJunWen
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

package owerr

// Error OneWay specific error.
// Switch based on code to handle specific error when using OneWayClient.
type Error interface {
	// Satisfy the generic error interface.
	error

	// Returns the error details message.
	Message() string

	// Returns the short phrase depicting the classification of the error.
	Code() string

	// Returns the status code of the HTTP response.
	StatusCode() int
}

// New initializes a new OneWayError.
func New(code, message string, statusCode int) Error {
	return newBaseError(code, message, statusCode)
}

const (
	// RequestFailure request Failure error. Error is thrown when response status is not OK.
	RequestFailure = "RequestFailure"

	// InvalidCredentials invalid Credentials error. Error is thrown when API Username or API Password is invalid.
	InvalidCredentials = "InvalidCredentials"

	// InvalidSenderID invalid Sender ID error. Error is thrown when Sender ID is not valid.
	InvalidSenderID = "InvalidSenderID"

	// InvalidMobileNo invalid Mobile Number error. Error is thrown when Mobile No is not valid.
	InvalidMobileNo = "InvalidMobileNo"

	// InvalidLanguageType invalid LanguageType error. Error is thrown when Language type is not valid.
	InvalidLanguageType = "InvalidLanguageType"

	// InvalidMessageCharacters invalid Message Characters error. Error is thrown when there are invalid characters in the request message.
	InvalidMessageCharacters = "InvalidMessageCharacters"

	// InsufficientCreditBalance insufficientCreditBalance error. Error is thrown when the user does not have sufficient credit balance to perform certain tasks.
	InsufficientCreditBalance = "InsufficientCreditBalance"

	// MTInvalidNotFound mobile terminating ID invalid or not found error. Error is thrown when Mobile Terminating ID is not a valid ID or not found.
	MTInvalidNotFound = "MTInvalidNotFound"

	// MessageDeliveryFailure message delivery failure error. Error is thrown when Message has been failed to deliver when calling check transaction status API.
	MessageDeliveryFailure = "MessageDeliveryFailure"

	// UnknownError unknown error. Unknown Response returned from OneWay API Gateway.
	UnknownError = "UnknownError"
)


owsms/client.go
// Copyright (c) KwanJunWen
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

package owsms

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"unicode/utf8"

	"github.com/junwen-k/onewaysms-sdk-go/owerr"
)

const version = "0.1.0"

// doer implements http.Client Do interface.
type doer interface {
	Do(req *http.Request) (*http.Response, error)
}

// Client OneWaySMS client structure.
// Based on specifications found in http://smsd2.onewaysms.sg/api.pdf.
type Client struct {
	client      doer
	baseURL     string
	apiUsername string
	apiPassword string
	senderID    string
}

// NewClient initializes a new OneWaySMS client.
func NewClient(baseURL, apiUsername, apiPassword, senderID string) *Client {
	return &Client{
		client:      http.DefaultClient,
		baseURL:     baseURL,
		apiUsername: apiUsername,
		apiPassword: apiPassword,
		senderID:    senderID,
	}
}

// NewClientWithHTTP initializes a new OneWaySMS client with custom http client.
func NewClientWithHTTP(baseURL, apiUsername, apiPassword, senderID string, client doer) *Client {
	c := NewClient(baseURL, apiUsername, apiPassword, senderID)
	c.client = client
	return c
}

func (c *Client) messageToHex(message string) string {
	buf := new(bytes.Buffer)
	for _, r := range message {
		buf.WriteString(strings.Trim(fmt.Sprintf("%U", r), "U+"))
	}
	return buf.String()
}

func (c *Client) getLanguageType(message string) LanguageType {
	m := message
	for len(m) > 0 {
		_, size := utf8.DecodeRuneInString(m)
		if size > 1 {
			return LanguageTypeUnicode
		}
		m = m[size:]
	}
	return LanguageTypeNormal
}

func (c *Client) buildRequestURL(path string, urlParams map[string]string) string {
	params := url.Values{}
	for k, v := range urlParams {
		params.Add(k, v)
	}
	return fmt.Sprintf("%s/%s?%s", c.baseURL, path, params.Encode())
}

func (c *Client) buildSendSMSRequestURL(input *SendSMSInput) string {
	if input.LanguageType == "" {
		input.LanguageType = c.getLanguageType(input.Message)
	}
	if input.LanguageType == LanguageTypeUnicode {
		input.Message = c.messageToHex(input.Message)
	}

	return c.buildRequestURL("api.aspx", map[string]string{
		"apiusername":  c.apiUsername,
		"apipassword":  c.apiPassword,
		"senderid":     c.senderID,
		"mobileno":     strings.Join(input.MobileNo, ","),
		"languagetype": string(input.LanguageType),
		"message":      input.Message,
	})
}

func (c *Client) buildCheckTransactionStatusRequestURL(input *CheckTransactionStatusInput) string {
	return c.buildRequestURL("bulktrx.aspx", map[string]string{
		"mtid": strconv.Itoa(input.MTID),
	})
}

func (c *Client) buildCheckCreditBalanceRequestURL() string {
	return c.buildRequestURL("bulkcredit.aspx", map[string]string{
		"apiusername": c.apiUsername,
		"apipassword": c.apiPassword,
	})
}

func (c *Client) getRequest(requestURL string) (*http.Response, error) {
	if c.client == nil {
		c.client = http.DefaultClient
	}

	req, err := http.NewRequest(http.MethodGet, requestURL, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", fmt.Sprintf("onewaysms-sdk-go/%s", version))

	return c.client.Do(req)
}

// SendSMS Initiate send SMS request. SMS's language type will be automatically set unless it is defined in the SMS request structure.
func (c *Client) SendSMS(input *SendSMSInput) (*SendSMSOutput, *http.Response, error) {
	requestURL := c.buildSendSMSRequestURL(input)

	resp, err := c.getRequest(requestURL)
	if err != nil {
		return nil, resp, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, resp, owerr.New(owerr.RequestFailure, "request failure", resp.StatusCode)
	}

	b, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, resp, err
	}

	mtIDs := make([]int, 0)
	for _, _mtID := range strings.Split(string(b), ",") {
		mtID, err := strconv.Atoi(strings.TrimSpace(_mtID))
		if err != nil {
			return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
		}
		mtIDs = append(mtIDs, mtID)
	}

	if len(mtIDs) <= 0 {
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}

	if mtIDs[0] > 0 {
		return &SendSMSOutput{MTIDs: mtIDs}, resp, nil
	}
	switch mtIDs[0] {
	case -100:
		return nil, resp, owerr.New(owerr.InvalidCredentials, "apiusername or apipassword is invalid", resp.StatusCode)
	case -200:
		return nil, resp, owerr.New(owerr.InvalidSenderID, "senderid parameter is invalid", resp.StatusCode)
	case -300:
		return nil, resp, owerr.New(owerr.InvalidMobileNo, "mobileno parameter is invalid", resp.StatusCode)
	case -400:
		return nil, resp, owerr.New(owerr.InvalidLanguageType, "languagetype is invalid", resp.StatusCode)
	case -500:
		return nil, resp, owerr.New(owerr.InvalidMessageCharacters, "characters in message are invalid", resp.StatusCode)
	case -600:
		return nil, resp, owerr.New(owerr.InsufficientCreditBalance, "insufficient credit balance", resp.StatusCode)
	default:
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}
}

// CheckTransactionStatus check transaction status based on mobile terminating ID provided.
func (c *Client) CheckTransactionStatus(input *CheckTransactionStatusInput) (*CheckTransactionStatusOutput, *http.Response, error) {
	requestURL := c.buildCheckTransactionStatusRequestURL(input)

	resp, err := c.getRequest(requestURL)
	if err != nil {
		return nil, resp, err
	}
	defer resp.Body.Close()

	b, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, resp, err
	}

	status, err := strconv.Atoi(strings.TrimSpace(string(b)))
	if err != nil {
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}

	switch status {
	case 0:
		return &CheckTransactionStatusOutput{Status: MTTransactionStatusSuccess}, resp, nil
	case 100:
		return &CheckTransactionStatusOutput{Status: MTTransactionStatusTelcoDelivered}, resp, nil
	case -100:
		return nil, resp, owerr.New(owerr.MTInvalidNotFound, "mtid is invalid or not found", resp.StatusCode)
	case -200:
		return nil, resp, owerr.New(owerr.MessageDeliveryFailure, "message delivery failed", resp.StatusCode)
	default:
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}
}

// CheckCreditBalance check remaining credit balance based on API Username and Password from client's config.
func (c *Client) CheckCreditBalance() (*CheckCreditBalanceOutput, *http.Response, error) {
	requestURL := c.buildCheckCreditBalanceRequestURL()

	resp, err := c.getRequest(requestURL)
	if err != nil {
		return nil, resp, err
	}
	defer resp.Body.Close()

	b, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, resp, err
	}

	creditBalance, err := strconv.ParseFloat(strings.TrimSpace(string(b)), 32)
	if err != nil {
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}

	if creditBalance >= 0 {
		return &CheckCreditBalanceOutput{CreditBalance: float32(creditBalance)}, resp, nil
	}

	switch creditBalance {
	case -100:
		return nil, resp, owerr.New(owerr.InvalidCredentials, "apiusername or apipassword is invalid", resp.StatusCode)
	default:
		return nil, resp, owerr.New(owerr.UnknownError, "unknown error", resp.StatusCode)
	}
}
